src/Faithlife.Testing/AssertEx.cs

```diff
@@ -842,6 +836,69 @@ private static string ToString(object obj)
 			return toString;
 		}
 
+		private static string ToPrettyJson(object obj, int tabLevel)
+		{
+			var rawJson = JsonUtility.ToJson(obj, s_jsonWithTabs).Replace("\r\n", "\n");
+			var lines = rawJson.Split('\n').Select(l => l.Trim()).ToList();
+
+			// if json can fit on one line, return inlined
+			var shortJson = lines.Select(l => l + (l.EndsWithOrdinal(",") ? " " : "")).Join("");
+			if (shortJson.Length < c_maxJsonLength)
+				return shortJson;
+
+			// else construct pretty json output
+			var longJson = "";
+			for (var i = 0; i < lines.Count; i++)
+			{
+				var line = lines[i];
+				var hasArrayStart = line.EndsWithOrdinal("[");
+				var hasStart = hasArrayStart || line.EndsWithOrdinal("{");
+				var hasEnd = line.StartsWithOrdinal("]") || line.StartsWithOrdinal("}");
+
+				if (hasArrayStart)
+				{
+					// try to inline the array if it's short and composed of primitives
+					var arrayLines = new List<string>();
+					var hasNested = false;
+					var j = i;
+					while (++j < lines.Count)
+					{
+						var arrayLine = lines[j];
+						if (arrayLine.StartsWithOrdinal("]"))
+							break;
+
+						if (arrayLine.StartsWithOrdinal("[") || arrayLine.StartsWithOrdinal("{"))
```

Comment:
On line 854 you use `line.EndsWithOrdinal("[")`; everywhere but here, you seem to be consistently using `StartsWith` for the closing-characeters `}` and `]` and `EndsWith` for the opening characters `[` and `{`.

Was that an intentional pattern, missed in one spot?

If `StartsWith` vs `EndsWith` doesn't matter, I'd advocate doing full-string matching on the one character we're interested in.

Reply:
So, here's the difference:
For *properties*, the start of an object or array will be at the *end* of the line (e.g. `Foo: {`), however, for objects or arrays that are *inside* arrays, the opening brace is at the start of the line. It might *also* be at the end, but in the case of an empty array or object, may not be. e.g.
```
Array: [
    {
        ...
    }
]

ArrayWithEmptyObject: [
    {}
]
```